### PR TITLE
8324678: Replace NULL with nullptr in HotSpot gtests

### PR DESCRIPTION
**Pull request by coleenp**
If this is sufficient, here's the change for NULL to nullptr, adjusting some obvious strings that had NULL in them maybe not all.
Tested with gtest,

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8324678](https://bugs.openjdk.org/browse/JDK-8324678): Replace NULL with nullptr in HotSpot gtests (**Enhancement** - P4)

### Reviewers
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to [d8e3d92d](https://git.openjdk.org/jdk/pull/17577/files/d8e3d92d2198cb9c92d1da923f5d1b2772edbd51)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [d8e3d92d](https://git.openjdk.org/jdk/pull/17577/files/d8e3d92d2198cb9c92d1da923f5d1b2772edbd51)
 * [Julian Waters](https://openjdk.org/census#jwaters) (@TheShermanTanker - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17577/head:pull/17577` \
`$ git checkout pull/17577`

Update a local copy of the PR: \
`$ git checkout pull/17577` \
`$ git pull https://git.openjdk.org/jdk.git pull/17577/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17577`

View PR using the GUI difftool: \
`$ git pr show -t 17577`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17577.diff">https://git.openjdk.org/jdk/pull/17577.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17577#issuecomment-1912634434)

### CODE REVIEW

File: test/hotspot/gtest/gc/shared/test_oopStorage_parperf.cpp [hunk 1]
```diff
@@ -152,7 +152,7 @@ class OopStorageParIterPerf::Task : public WorkerTask {
 
 class OopStorageParIterPerf::Closure : public OopClosure {
 public:
-  virtual void do_oop(oop* p) { guarantee(*p == NULL, "expected NULL"); }
+  virtual void do_oop(oop* p) { guarantee(*p == nullptr, "expected nullptr"); }
```
Comment (kimbarrett) [inline test/hotspot/gtest/gc/shared/test_oopStorage_parperf.cpp:155]: s/expected nullptr/expected null/

File: test/hotspot/gtest/gc/shared/test_collectedHeap.cpp [hunk 2]
```diff
@@ -32,8 +32,8 @@ TEST_VM(CollectedHeap, is_in) {
   uintptr_t epsilon = (uintptr_t) MinObjAlignment;
   uintptr_t outside_heap = (uintptr_t) &epsilon;
 
-  // Test that NULL is not in the heap.
-  ASSERT_FALSE(heap->is_in(NULL)) << "NULL is unexpectedly in the heap";
+  // Test that nullptr is not in the heap.
+  ASSERT_FALSE(heap->is_in(nullptr)) << "nullptr is unexpectedly in the heap";
```
Comment (kimbarrett) [inline test/hotspot/gtest/gc/shared/test_collectedHeap.cpp:36]: s/"nullptr is/"null is/

File: test/hotspot/gtest/memory/test_guardedMemory.cpp [hunk 3]
```diff
@@ -30,7 +30,7 @@
 #define GEN_PURPOSE_TAG ((void *) ((uintptr_t)0xf000f000))
 
 static void guarded_memory_test_check(void* p, size_t sz, void* tag) {
-  ASSERT_TRUE(p != NULL) << "NULL pointer given to check";
+  ASSERT_TRUE(p != nullptr) << "nullptr pointer given to check";
```
Comment (kimbarrett) [inline test/hotspot/gtest/memory/test_guardedMemory.cpp:33]: s/nullptr pointer/null pointer/

File: test/hotspot/gtest/nmt/test_nmtpreinit.cpp [hunk 4]
```diff
@@ -84,11 +84,11 @@ class TestAllocations {
     os::free(os_malloc(0));              // 0-sized allocation, should be free-able
     p2 = os_realloc(os_malloc(10), 20);  // realloc, growing
     p3 = os_realloc(os_malloc(20), 10);  // realloc, shrinking
-    p4 = os_realloc(NULL, 10);           // realloc with NULL pointer
+    p4 = os_realloc(nullptr, 10);           // realloc with nullptr pointer
```
Comment (kimbarrett) [inline test/hotspot/gtest/nmt/test_nmtpreinit.cpp:87]: s/nullptr pointer/null pointer/

File: test/hotspot/gtest/runtime/test_ThreadsListHandle.cpp [hunk 5]
```diff
@@ -213,18 +213,18 @@ TEST_VM(ThreadsListHandle, sanity) {
 
     // Test case: after first nested ThreadsListHandle (tlh2) has been destroyed
 
-    // Verify the current thread's hazard ptr is NULL:
-    EXPECT_EQ(ThreadsListHandleTest::get_Thread_threads_hazard_ptr(thr), (ThreadsList*)NULL)
-        << "thr->_threads_hazard_ptr must be NULL";
+    // Verify the current thread's hazard ptr is nullptr:
```
Comment (kimbarrett) [inline test/hotspot/gtest/runtime/test_ThreadsListHandle.cpp:216]: s/nullptr/null/

File: test/hotspot/gtest/runtime/test_ThreadsListHandle.cpp [hunk 6]
```diff
@@ -397,9 +397,9 @@ TEST_VM(ThreadsListHandle, sanity) {
 
       // Test case: after double nested ThreadsListHandle (tlh3) has been destroyed
 
-      // Verify the current thread's hazard ptr is NULL:
-      EXPECT_EQ(ThreadsListHandleTest::get_Thread_threads_hazard_ptr(thr), (ThreadsList*)NULL)
-          << "thr->_threads_hazard_ptr must be NULL";
+      // Verify the current thread's hazard ptr is nullptr:
```
Comment (kimbarrett) [inline test/hotspot/gtest/runtime/test_ThreadsListHandle.cpp:400]: s/nullptr/null/

File: test/hotspot/gtest/runtime/test_ThreadsListHandle.cpp [hunk 7]
```diff
@@ -445,18 +445,18 @@ TEST_VM(ThreadsListHandle, sanity) {
 
     // Test case: after first nested ThreadsListHandle (tlh2) has been destroyed
 
-    // Verify the current thread's hazard ptr is NULL:
-    EXPECT_EQ(ThreadsListHandleTest::get_Thread_threads_hazard_ptr(thr), (ThreadsList*)NULL)
-        << "thr->_threads_hazard_ptr must be NULL";
+    // Verify the current thread's hazard ptr is nullptr:
```
Comment (kimbarrett) [inline test/hotspot/gtest/runtime/test_ThreadsListHandle.cpp:448]: s/nullptr/null/

File: test/hotspot/gtest/runtime/test_ThreadsListHandle.cpp [hunk 8]
```diff
@@ -565,18 +565,18 @@ TEST_VM(ThreadsListHandle, sanity) {
 
     // Test case: after first back-to-back nested ThreadsListHandle (tlh2a) has been destroyed
 
-    // Verify the current thread's hazard ptr is NULL:
-    EXPECT_EQ(ThreadsListHandleTest::get_Thread_threads_hazard_ptr(thr), (ThreadsList*)NULL)
-        << "thr->_threads_hazard_ptr must be NULL";
+    // Verify the current thread's hazard ptr is nullptr:
```
Comment (kimbarrett) [inline test/hotspot/gtest/runtime/test_ThreadsListHandle.cpp:568]: s/nullptr/null/

File: test/hotspot/gtest/runtime/test_ThreadsListHandle.cpp [hunk 9]
```diff
@@ -643,18 +643,18 @@ TEST_VM(ThreadsListHandle, sanity) {
 
     // Test case: after second back-to-back nested ThreadsListHandle (tlh2b) has been destroyed
 
-    // Verify the current thread's hazard ptr is NULL:
-    EXPECT_EQ(ThreadsListHandleTest::get_Thread_threads_hazard_ptr(thr), (ThreadsList*)NULL)
-        << "thr->_threads_hazard_ptr must be NULL";
+    // Verify the current thread's hazard ptr is nullptr:
```
Comment (kimbarrett) [inline test/hotspot/gtest/runtime/test_ThreadsListHandle.cpp:646]: s/nullptr/null/

File: test/hotspot/gtest/runtime/test_classLoader.cpp [hunk 10]
```diff
@@ -30,32 +30,32 @@
 // Tests ClassLoader::package_from_class_name()
 TEST_VM(ClassLoader, null_class_name) {
   bool bad_class_name = false;
-  TempNewSymbol retval = ClassLoader::package_from_class_name(NULL, &bad_class_name);
-  ASSERT_TRUE(bad_class_name) << "Function did not set bad_class_name with NULL class name";
-  ASSERT_TRUE(retval == NULL) << "Wrong package for NULL class name pointer";
+  TempNewSymbol retval = ClassLoader::package_from_class_name(nullptr, &bad_class_name);
+  ASSERT_TRUE(bad_class_name) << "Function did not set bad_class_name with nullptr class name";
```
Comment (kimbarrett) [inline test/hotspot/gtest/runtime/test_classLoader.cpp:34]: s/nullptr/null/

File: test/hotspot/gtest/runtime/test_classLoader.cpp [hunk 11]
```diff
@@ -30,32 +30,32 @@
...
+  ASSERT_TRUE(retval == nullptr) << "Wrong package for nullptr class name pointer";
```
Comment (kimbarrett) [inline test/hotspot/gtest/runtime/test_classLoader.cpp:35]: second s/nullptr/null/

File: test/hotspot/gtest/runtime/test_os_linux.cpp [hunk 12]
```diff
@@ -200,8 +200,8 @@ TEST_VM(os_linux, reserve_memory_special_huge_tlbfs_size_not_aligned_with_bad_re
       char* p = HugeTlbfsMemory::reserve_memory_special_huge_tlbfs(size, alignment, lp, req_addr, false);
       HugeTlbfsMemory mr(p, size);
       // as the area around req_addr contains already existing mappings, the API should always
-      // return NULL (as per contract, it cannot return another address)
-      EXPECT_TRUE(p == NULL) << " size = " << size
+      // return nullptr (as per contract, it cannot return another address)
```
Comment (kimbarrett) [inline test/hotspot/gtest/runtime/test_os_linux.cpp:203]: s/nullptr/null/

File: test/hotspot/gtest/runtime/test_os_linux_cgroups.cpp [hunk 13]
```diff
@@ -63,13 +63,13 @@ TEST(cgroupTest, set_cgroupv1_subsystem_path) {
 TEST(cgroupTest, set_cgroupv2_subsystem_path) {
   TestCase at_mount_root = {
     "/sys/fs/cgroup",       // mount_path
-    NULL,                   // root_path, ignored
+    nullptr,                   // root_path, ignored
```
Comment (kimbarrett) [inline test/hotspot/gtest/runtime/test_os_linux_cgroups.cpp:66]: comment mis-indented

File: test/hotspot/gtest/runtime/test_os_linux_cgroups.cpp [hunk 14]
```diff
@@ -63,13 +63,13 @@ TEST(cgroupTest, set_cgroupv1_subsystem_path) {
...
     "/",                    // cgroup_path
     "/sys/fs/cgroup"        // expected_path
   };
   TestCase sub_path = {
     "/sys/fs/cgroup",       // mount_path
-    NULL,                   // root_path, ignored
+    nullptr,                   // root_path, ignored
```
Comment (kimbarrett) [inline test/hotspot/gtest/runtime/test_os_linux_cgroups.cpp:72]: comment mis-indented

File: test/hotspot/gtest/testutils.hpp [hunk 15]
```diff
@@ -53,8 +53,8 @@ class GtestUtils : public AllStatic {
 #define EXPECT_RANGE_IS_MARKED(p, size)             EXPECT_TRUE(GtestUtils::is_range_marked(p, size))
 
 // Mimicking the official ASSERT_xx and EXPECT_xx counterparts of the googletest suite.
-// (ASSERT|EXPECT)_NOT_NULL: check that the given pointer is not NULL
-// (ASSERT|EXPECT)_NULL: check that the given pointer is NULL
+// (ASSERT|EXPECT)_NOT_NULL: check that the given pointer is not nullptr
+// (ASSERT|EXPECT)_NULL: check that the given pointer is nullptr
```
Comment (kimbarrett) [inline test/hotspot/gtest/testutils.hpp:57]: both s/nullptr/null/

**Review by kimbarrett — CHANGES_REQUESTED** [review]
Mostly good.  A few nullptr -> null, and a couple other nits.

File: test/hotspot/gtest/runtime/test_os_linux.cpp (hunk 12 above)
Comment (coleenp) [reply test/hotspot/gtest/runtime/test_os_linux.cpp:203]: This is technically correct.  Changing the comments like this seems really pointless.

**Review by kimbarrett — APPROVED** [review]
Looks good.

File: test/hotspot/gtest/nmt/test_nmt_buffer_overflow_detection.cpp [hunk 16]
```diff
@@ -109,7 +109,7 @@ DEFINE_TEST(test_double_free, "header canary")
 ///////
 
 static void test_invalid_block_address() {
-  // very low, like the result of an overflow or of accessing a NULL this pointer
+  // very low, like the result of an overflow or of accessing a nullptr this pointer
```
Comment (dholmes-ora) [inline test/hotspot/gtest/nmt/test_nmt_buffer_overflow_detection.cpp:112]: s/nullptr/null

This is referring to nullness as a concept.

**Review by dholmes-ora — APPROVED** [review]
Looks good.

Part of the cleanup with using `nullptr` is that we should never need to cast `nullptr` to another pointer type. I flagged a few cases below but certainly not all.

For text/comments the general rule is that when talking about the general concept of nullness then we can say "null", whereas if referring to an actual code artifact we use "nullptr". Sometimes either will be fine e.g "returns null on error", "returns nullptr on error".

Thanks.

**Review by TheShermanTanker — APPROVED** [review]

**Comment by kimbarrett** [comment]
> Part of the cleanup with using `nullptr` is that we should never need to cast `nullptr` to another pointer type. I flagged a few cases below but certainly not all.

I think this may eventually be true, but I don't think it is currently true.  I think we will need to add overloads for
std::nullptr_t in some places in order to remove some casts.  This is particularly true for templates, and I think all
of the places you mentioned are of this kind.

File: test/hotspot/gtest/metaspace/metaspaceGtestCommon.hpp [hunk 17]
```diff
@@ -145,12 +145,12 @@ void mark_range(MetaWord* p, size_t word_size);
 void check_marked_range(const MetaWord* p, size_t word_size);
 
 //////////////////////////////////////////////////////////
-// Some helpers to avoid typing out those annoying casts for NULL
+// Some helpers to avoid typing out those annoying casts for nullptr
```
[the remark this answers is not in the document]
Comment (coleenp) [reply test/hotspot/gtest/metaspace/metaspaceGtestCommon.hpp:148]: Like I said in the description, we can change the comments on a case-by-case basis in order to get the bulk of the change done mechanically.

[the remark this answers is not in the document]
Comment (coleenp) [reply test/hotspot/gtest/metaspace/metaspaceGtestCommon.hpp:148]: ie, we should file a separate RFE for comments.

[the remark this answers is not in the document]
Comment (coleenp) [reply test/hotspot/gtest/metaspace/metaspaceGtestCommon.hpp:148]: And in this comment nullptr makes perfect sense.

**Comment by coleenp** [comment]
Thanks Julian, Kim and David for the reviews.  I fixed the null casts that you pointed out and the one comment, and fixed my copyright script (so updated the copyrights).  I'll wait for GHA to integrate.

**Review by TheShermanTanker — APPROVED** [review]

File: test/hotspot/gtest/metaspace/metaspaceGtestCommon.hpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply test/hotspot/gtest/metaspace/metaspaceGtestCommon.hpp:148]: Yeah I wasn't commenting on the use of nullptr in the comment :)

**Pull request closed without merging**